### PR TITLE
refactor: core module & compose set up

### PR DESCRIPTION
**Pull request by PratyushSingh07**
- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

**Comment by PratyushSingh07** [comment]
Hi @jawidMuhammadi , i require your review on this PR , this may not contain everything that we had discussed as I have sadly forgotten some of the points. 
Could you please take a look at this and review so that I can iterate on them

File: app/build.gradle [hunk 1]
```diff
@@ -183,5 +192,13 @@ dependencies {
     implementation("com.google.dagger:hilt-android:2.44")
     kapt("com.google.dagger:hilt-android-compiler:2.47")
 
+    // Jetpack Compose
+    implementation "androidx.compose.material:material:$rootProject.composeVersion"
```
Comment (jawidMuhammadi) [inline app/build.gradle:199]: Can we use Comopse BoM? -> https://developer.android.com/jetpack/compose/bom?authuser=1

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsFragment.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        return ComposeView(requireContext()).apply {
+            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+            setContent {
+                AboutUsScreen()
+            }
+        }
+    }
+
+    companion object {
+        @JvmStatic
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsFragment.kt]: Remove JvmStatic, the whole project is in Kotlin so we don't need it

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        fun newInstance(): AboutUsFragment {
+            val fragment = AboutUsFragment()
+            val args = Bundle()
+            fragment.arguments = args
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsFragment.kt]: I belive there is no need for passing args, you can remove it.

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import org.mifos.mobile.utils.StartActivity
+import java.util.*
+
+@SuppressWarnings("UnusedMaterial3ScaffoldPaddingParameter")
+@Composable
+fun AboutUsScreen(viewModel: AboutUsViewModel = viewModel()) {
+    val context = StartActivity(LocalContext.current)
+
+    MifosMobileTheme {
+        Scaffold {
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: Use the MifosMobileThem and Scaffold to wrap the screen level composable, which means, 
it should be used inside the fragment to wrap the `AboutUsScreen` composable. Remove it from here.

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import androidx.lifecycle.ViewModel
+import dagger.hilt.android.lifecycle.HiltViewModel
+import org.mifos.mobile.R
+import org.mifos.mobile.models.AboutUsItem
+import javax.inject.Inject
+
+@HiltViewModel
+class AboutUsViewModel @Inject constructor() : ViewModel() {
+
+    val aboutUsList: List<AboutUsItem> = listOf(
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsViewModel.kt]: Rename it to `aboutUseItems`

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                    modifier = Modifier
+                        .fillMaxWidth()
+                        .padding(bottom = 16.dp)
+                )
+
+                LazyColumn {
+                    items(viewModel.aboutUsList) { item ->
+                        MifosItemCard(onClick = {
+                            when (item.name) {
+                                R.string.official_website -> {
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: As we have discussed, it is not a good idea to check with strings inside the composable.
Instead, you create an enum class as follows:

```
  enum class AboutUsListItemId{
        OfficialWebsite,
        PrivacyPolicy,
        ...
    }
```

Then use this this as a param in `AboutUsItem` data class:

```
data class AboutUsItem(
    val name: Int,
    val iconUrl: Int? = null,
    val itemId: AboutUsListItemId
)
```

Then do the checks based on itemId

File: app/src/main/java/org/mifos/mobile/utils/StartActivity.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package org.mifos.mobile.utils
+
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+
+class StartActivity(private val context: Context) {
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/utils/StartActivity.kt]: We don't need this class, it is just over-engineering 😊

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                                    context.startActivity(WEBSITE_LINK)
+                                }
+                                R.string.licenses -> {
+                                    context.startActivity(LICENSE_LINK)
+                                }
+                                R.string.privacy_policy -> {
+                                    context.startActivity(PrivacyPolicyActivity::class.java)
+                                }
+                                R.string.sources -> {
+                                    context.startActivity(SOURCE_CODE_LINK)
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: Create another variable inside the ViewModel named `AboutUsItemEvent`, you can use a liveData and StateFlow and then observe it inside the fragment. Based on the itemId navigate to the proper screens. I mean the navigation should be done inside the fragment not this composable.

**Review by jawidMuhammadi — CHANGES_REQUESTED** [review]

File: app/build.gradle (hunk 1 above)
Comment (PratyushSingh07) [reply app/build.gradle:199]: Actually  can we skip it for the time being ?  I tried to implement this and when I pushed the code I got the below error from the ci cd 
image

To resolve this issue , I would have to check every dependency individually just to make sure that BOM provides it. Even if i do all of that I  will get the below error in `build.gradle(:core)`
image

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt (hunk 6 above)
Comment (PratyushSingh07) [reply app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: done

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt (hunk 8 above)
Comment (PratyushSingh07) [reply app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: done

**Comment by PratyushSingh07** [comment]
Done the changes @jawidMuhammadi , do recview them

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+        Text(
+            text = stringResource(id = R.string.about_app_description),
+            style = MaterialTheme.typography.titleSmall.copy(),
+            textAlign = TextAlign.Center,
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(bottom = 16.dp)
+        )
+
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: These three components can be moved to a separate composable. You can name the composable something like `AboutUsHeader`.

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+@Composable
+fun AboutUsItemCard(
+    name: Int,
+    iconUrl: Int? = null,
+) {
+    val currentYear = Calendar.getInstance().get(Calendar.YEAR)
+    val copyrightText =
+        LocalContext.current.getString(R.string.copy_right_mifos)
+            .replace("%1\$s", currentYear.toString())
+
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: These logics need to be handled in ViewModel and passed as name/label

File: core/src/main/java/org/mifos/mobile/core/ui/component/MifosItemComponent.kt [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.unit.dp
+
+@Composable
+fun MifosItemContent(name: Int, iconUrl: Int?) {
```
Comment (jawidMuhammadi) [inline core/src/main/java/org/mifos/mobile/core/ui/component/MifosItemComponent.kt]: Rename to `AboutUseItemCard`

File: app/src/main/java/org/mifos/mobile/models/AboutUsItem.kt [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package org.mifos.mobile.models
+
+import org.mifos.mobile.ui.enums.AboutUsListItemId
+
+data class AboutUsItem(
+    val name: Int,
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/models/AboutUsItem.kt]: Rename this to the `title` and add another item named `subtitle` (You would need this for the copyright item )

File: core/src/main/java/org/mifos/mobile/core/ui/component/CopyrightItemContent.kt [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+import androidx.compose.foundation.layout.padding
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.unit.dp
+
+@Composable
+fun CopyrightItemContent(name: Int, copyrightText: String) {
```
Comment (jawidMuhammadi) [inline core/src/main/java/org/mifos/mobile/core/ui/component/CopyrightItemContent.kt]: You don't need to create this Composable, the `AboutUsItemCard` should handle this cases where there is a title and `subtitle`

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    if (name != R.string.license_string_with_value) {
+        MifosItemContent(name = name, iconUrl = iconUrl)
+    } else {
+        CopyrightItemContent(name = name, copyrightText = copyrightText)
+    }
+}
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: We can remove this logic

File: core/src/main/java/org/mifos/mobile/core/ui/component/MifosItemCard.kt [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    shape: Shape = RoundedCornerShape(8.dp),
+    elevation: Dp = 1.dp,
+    onClick: () -> Unit,
+    content: @Composable ColumnScope.() -> Unit
+) {
+    Card(
+        shape = shape,
+        modifier = modifier
+            .fillMaxWidth()
+            .padding(bottom = 8.dp)
```
Comment (jawidMuhammadi) [inline core/src/main/java/org/mifos/mobile/core/ui/component/MifosItemCard.kt]: Remove the padding from the here

**Review by jawidMuhammadi — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt (hunk 10 above)
Comment (PratyushSingh07) [reply app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: done

File: core/src/main/java/org/mifos/mobile/core/ui/component/MifosItemComponent.kt (hunk 11 above)
Comment (PratyushSingh07) [reply core/src/main/java/org/mifos/mobile/core/ui/component/MifosItemComponent.kt]: done

File: app/src/main/java/org/mifos/mobile/models/AboutUsItem.kt (hunk 12 above)
Comment (PratyushSingh07) [reply app/src/main/java/org/mifos/mobile/models/AboutUsItem.kt]: done

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt (hunk 14 above)
Comment (PratyushSingh07) [reply app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: done

**Comment by PratyushSingh07** [comment]
can you please review it once more @jawidMuhammadi  !!

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsViewModel.kt [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+    private val currentYear = Calendar.getInstance().get(Calendar.YEAR)
+    private val copyrightText =
+        context?.getString(R.string.copy_right_mifos)
+            ?.replace("%1\$s", currentYear.toString())
+
+    val aboutUsItems: List<AboutUsItem> = listOf(
+        AboutUsItem(
+            context?.getString(R.string.app_version_text),
+            null,
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsViewModel.kt]: You don't need to pass `null` value here, the default value is already `null`

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+import org.mifos.mobile.ui.enums.AboutUsListItemId
+import java.util.*
+
+@SuppressWarnings("UnusedMaterial3ScaffoldPaddingParameter")
+@Composable
+fun AboutUsScreen(viewModel: AboutUsViewModel = viewModel()) {
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+            .padding(16.dp, 56.dp, 16.dp, 16.dp)
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: You can remove this `56dp` and use one padding `16dp`, and you can use the Spacer widget to add space

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsFragment.kt [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View {
+        return ComposeView(requireContext()).apply {
+            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+            setContent {
+                MifosMobileTheme {
+                    Scaffold {
+                        AboutUsScreen()
```
Comment (jawidMuhammadi) [inline app/src/main/java/org/mifos/mobile/ui/about/AboutUsFragment.kt]: Pass the viewModel instance in `AboutScreen`, you should use the same instance of the view model in both fragment and the compose screen

File: core/src/main/java/org/mifos/mobile/core/ui/component/AboutUsItemCard.kt [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+@Composable
+fun AboutUsItemCard(title: String, subtitle: Int?, iconUrl: Int?) {
+    Row(
+        modifier = Modifier.padding(16.dp)
+    ) {
+        iconUrl?.let { painterResource(id = it) }?.let {
+            Image(
+                painter = it,
+                contentDescription = null,
+                modifier = Modifier.padding(0.dp, 0.dp, 10.dp, 0.dp)
```
Comment (jawidMuhammadi) [inline core/src/main/java/org/mifos/mobile/core/ui/component/AboutUsItemCard.kt]: The default value is already `0.dp`, you don't need to pass it.

File: core/src/main/java/org/mifos/mobile/core/ui/component/AboutUsItemCard.kt [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+            )
+        }
+        Column {
+            Text(
+                text = title,
+                style = MaterialTheme.typography.bodyLarge,
+                modifier = Modifier.padding(0.dp, 0.dp, 10.dp, 0.dp)
```
Comment (jawidMuhammadi) [inline core/src/main/java/org/mifos/mobile/core/ui/component/AboutUsItemCard.kt]: Applies same here

File: core/src/main/java/org/mifos/mobile/core/ui/component/AboutUsItemCard.kt [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+            )
+            if (subtitle != null) {
+                Text(
+                    text = stringResource(id = subtitle),
+                    style = MaterialTheme.typography.bodyLarge,
+                    modifier = Modifier.padding(0.dp, 0.dp, 10.dp, 0.dp)
```
Comment (jawidMuhammadi) [inline core/src/main/java/org/mifos/mobile/core/ui/component/AboutUsItemCard.kt]: For spacing and paddings use material design recommended values which are (8, 12, 16, 24)

**Review by jawidMuhammadi — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt (hunk 17 above)
Comment (PratyushSingh07) [reply app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: I will try to change it but the xml version had 56 dp as the padding and thats what i have used here

Comment (PratyushSingh07) [reply app/src/main/java/org/mifos/mobile/ui/about/AboutUsScreen.kt]: done

File: app/src/main/java/org/mifos/mobile/ui/about/AboutUsViewModel.kt (hunk 16 above)
Comment (PratyushSingh07) [reply app/src/main/java/org/mifos/mobile/ui/about/AboutUsViewModel.kt]: done

File: core/src/main/java/org/mifos/mobile/core/ui/component/AboutUsItemCard.kt (hunk 21 above)
Comment (PratyushSingh07) [reply core/src/main/java/org/mifos/mobile/core/ui/component/AboutUsItemCard.kt]: I had to skip this review as the rows had started to look bloated. I went with the values of 8.dp instead.

**Comment by PratyushSingh07** [comment]
Do review this @jawidMuhammadi , I would be over the moon if we could finish this task before 28th . Please ask for changes wherever required !!

**Review by jawidMuhammadi — APPROVED** [review]
LGTM, great job! 👍

**Pull request merged**